gcovr/gcov_parser.py

```diff
@@ -0,0 +1,994 @@
+    >>> _parse_line('branch 3 taken 15%')
+    _BranchLine(branchno=3, hits=1, annotation=None)
+    >>> _parse_line('branch 3 taken 0%')
+    _BranchLine(branchno=3, hits=0, annotation=None)
+    >>> _parse_line('branch 3 taken 123')
+    _BranchLine(branchno=3, hits=123, annotation=None)
+    >>> _parse_line('branch 7 taken 3% (fallthrough)')
+    _BranchLine(branchno=7, hits=1, annotation='fallthrough')
+    >>> _parse_line('branch 17 taken 99% (throw)')
+    _BranchLine(branchno=17, hits=1, annotation='throw')
+    >>> _parse_line('branch  0 never executed')
+    _BranchLine(branchno=0, hits=0, annotation=None)
+
+    Example: can parse call tags:
+    >>> _parse_line('call  0 never executed')
+    _CallLine(callno=0, returned=0)
+    >>> _parse_line('call  17 returned 50%')
+    _CallLine(callno=17, returned=1)
+    >>> _parse_line('call  17 returned 9')
+    _CallLine(callno=17, returned=9)
+
+    Example: can parse unconditional branches
+    >>> _parse_line('unconditional 1 taken 17')
+    _UnconditionalLine(branchno=1, hits=17)
+
+    Example: can parse function tags:
+    >>> _parse_line('function foo called 2 returned 95% blocks executed 85%')
+    _FunctionLine(name='foo', calls=2, returned=1, blocks_covered=1)
+
+    Example: can parse template specialization markers:
+    >>> _parse_line('------------------')
+    _SpecializationMarkerLine()
+
+    Example: can parse template specialization names:
+    >>> _parse_line('Foo<bar>::baz():')
+    _SpecializationNameLine(name='Foo<bar>::baz()')
+
+    Example: can parse block line:
+    >>> _parse_line('     1: 32-block  0')
+    _BlockLine(hits=1, lineno=32, blockno=0, extra_info=_ExtraInfo.NONE)
+    >>> _parse_line(' %%%%%: 33-block  1')
+    _BlockLine(hits=0, lineno=33, blockno=1, extra_info=_ExtraInfo.NONE)
+    >>> _parse_line(' $$$$$: 33-block  1')
+    _BlockLine(hits=0, lineno=33, blockno=1, extra_info=_ExtraInfo.EXCEPTION_ONLY)
+
+    Example: will reject garbage:
+    >>> _parse_line('nonexistent_tag foo bar')
+    Traceback (most recent call last):
+    gcovr.gcov_parser.UnknownLineType: nonexistent_tag foo bar
+    """
+    # pylint: disable=too-many-branches
+
+    tag = _parse_tag_line(line)
+    if tag is not None:
+        return tag
+
+    # Handle lines that are like source lines.
+    # But this could also include metadata lines and block-coverage lines.
+
+    # CODE
+    #
+    # Structure: "COUNT: LINENO:CODE"
+    #
+    # Examples:
+    #     -: 13:struct Foo{};
+    #    12: 13:foo += 1;
+    # #####: 13:foo += 1;
+    # =====: 13:foo += 1;
+    #   12*: 13:cond ? bar() : baz();
+    segments = line.split(":", 2)
+    if len(segments) == 3:
+        count_str, lineno, source_code = segments
+        count_str = count_str.strip()
+        lineno = lineno.strip()
+
+        # METADATA (key, value)
+        if count_str == "-" and lineno == "0":
+            key, value = source_code.split(":", 1)
+            return _MetadataLine(key.lower(), value)
```

Comment:
Why do you change the key to lowercase? I prefer keeping the original case.

Reply:
I did that because the upstream gcov format is subjectively “wrong”, but you're right that keeping this consistent is better. Done!